### PR TITLE
SiFive HiFive (FE310) RISC-V support

### PR DESCRIPTION
**Pull request by dgarske**
* HiFive1 HAL Support for PLL Clock, UART, RTC and Flash QSPI Erase/Write.
* HiFive1 update demo application for accepting firmware updates over UART.
* Added test-update-server application for pushing firmware image over UART.
* Fixes for building with `make SIGN=ECC256`.
* Improvements to wolfCrypt `user_settings.h`.
* General library cleanup (license headers and formatting)
* Updated the wolfSSL submodule to latest.
* Documentation updates including new `Targets.md` section for hardare instructions.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -67,3 +67,13 @@ cscope.out
...
 
 
+image.bin
+image_v1_signed.bin
+.wolfboot-arch-offset
+.wolfboot-offset
+.wolfboot-partition-size
+factory.bin
+wolfboot-align.bin
+wolfboot.bin
```
Comment (danielinux) [inline .gitignore]: Let's add *.bin

File: docs/Targets.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+### Stock bootloader
+Start Address: 0x20000000 is 64KB. Provides a "double tap" reset feature to halt boot and allow debugger to attach for reprogramming. Press reset button, when green light comes on press reset button again, then board will flash red.
+
+### Application Code
+Start Address: 0x20010000
+
+### wolfBoot configuration
+
+The default wolfBoot configuration will add a second stage bootloader, leaving the stock "double tap" bootloader for safety.
```
Comment (danielinux) [inline docs/Targets.md]: 'for safety' ... it's not clear to me how leaving the doubletap as stage1 increases safety

File: docs/Targets.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    If using the `riscv64-unknown-elf-` cross compiler you can add `CROSS_COMPILE=riscv64-unknown-elf-` to your `make` or modify `arch.mk` as follows:
+    
+    ```
+     ifeq ($(ARCH),RISCV)
+    -  CROSS_COMPILE:=riscv32-unknown-elf-
+    +  CROSS_COMPILE:=riscv64-unknown-elf-
+    ```
+
+
+2. `include/target.h`
```
Comment (danielinux) [inline docs/Targets.md]: Review addresses as discussed f2f

File: arch.mk [hunk 4]
```diff
@@ -2,18 +2,20 @@
 
 # check for FASTMATH or SP_MATH
 ifeq ($(SPMATH),1)
-  MATH_OBJS:=./lib/wolfssl/wolfcrypt/src/sp_int.o
+  MATH_OBJS:=./lib/wolfssl/wolfcrypt/src/sp_int.o ./lib/wolfssl/wolfcrypt/src/sp_c32.o
```
Comment (danielinux) [inline arch.mk]: This breaks compilation with `SIGN=ECC256 NO_ASM=1` on ARM (duplicate object). Also, we should not include sp_c32.o when compiling with assembly optimization (i.e. `NO_ASM=0`, which is the default for ECC). If you want to add sp_c32.o for RISC-V, please put it under `ifeq ($(ARCH),RISCV)`

File: docs/Targets.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+Bootloader Size: 0x10000 (64KB)
+
+```c
+#define WOLFBOOT_SECTOR_SIZE                 0x10000
+#define WOLFBOOT_PARTITION_BOOT_ADDRESS      0x20010000
```
Comment (danielinux) [inline docs/Targets.md]: This is the partition address for the main BOOT image, 0x100 before the application entry point

File: docs/Targets.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+```
+
+Application Size 0x40000 (256KB)
+
+```c
+#define WOLFBOOT_PARTITION_SIZE              0x40000
+#define WOLFBOOT_PARTITION_UPDATE_ADDRESS    0x20020000
```
Comment (danielinux) [inline docs/Targets.md]: this is the address of the UPDATE partition where the application can store the new firmware received during the update

File: docs/Targets.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+`JLinkGDBServer -device FE310 -port 3333`
+
+In another terminal:
+`riscv64-unknown-elf-gdb test-app/image.elf -ex "set remotetimeout 240" -ex "target extended-remote localhost:3333"`
+or
+`riscv64-unknown-elf-gdb wolfboot.elf -ex "set remotetimeout 240" -ex "target extended-remote localhost:3333"`
+
+
+
+## STM32-F407
```
Comment (danielinux) [inline docs/Targets.md]: This should be the first target described in Targets.md since it's the default (CI)

File: include/target.h [hunk 8]
```diff
@@ -1,29 +1,74 @@
...
 
 #ifdef EXT_FLASH
 
 /* Test configuration with 1MB external memory */
 /* (Addresses are relative to the beginning of the ext)*/
 
-#   define WOLFBOOT_PARTITION_SIZE			    0x80000
-#   define WOLFBOOT_PARTITION_UPDATE_ADDRESS    0x00000
-#   define WOLFBOOT_PARTITION_SWAP_ADDRESS      0x80000
+#ifndef WOLFBOOT_PARTITION_SIZE
```
Comment (danielinux) [inline include/target.h]: Not sure these ifndef are useful: target.h must be included everywhere and those symbols cannot be defined elsewhere.

File: docs/Targets.md (hunk 2 above)
Comment (danielinux) [reply docs/Targets.md]: also, from the addresses seems like you are actually replacing it

**Comment by danielinux** [comment]
CI tests passing, no regressions.

One known bug left in test-app/hifive.c (long jump/memory symbols)

**Comment by danielinux** [comment]
@dgarske could you please review my latest changes? eSPI driver works in the test-application as well now.

**Comment by dgarske** [comment]
PR Approved! Thanks Daniele!

**Pull request merged**